### PR TITLE
Add assorted list functions to prelude

### PR DESCRIPTION
**Pull request by gregorybchris**
Adding the following functions to the prelude:
- quicksort
- concat
- map
- range
- foldr
- take
- all
- any

### CODE REVIEW

**Comment by gregorybchris** [comment]
@tekknolagi I know at first you were excited about adding quicksort to the prelude. Perhaps this large addition of multiple functions is more than you bargained for though..?

If there's a concern that this beautifully simple language is getting too complex, maybe we want to remove the prelude altogether? If we don't want map and reduce, the same probably apples to filter and id (which already exist in the prelude).

Alternatively, we could avoid polluting the global namespace too much by making the prelude opt-in via a CLI flag. Then we can continue to add common functions to the prelude without affecting all users of the interpreter. Something like `python scrapscript.py eval program.scrap --prelude`. I would personally prefer this option, at least until we have scrapyards to pull in dependencies.

**Comment by tekknolagi** [comment]
I think we can just dump them all in for now. We can namespace later.

File: scrapscript.py [hunk 1]
```diff
@@ -3279,6 +3279,262 @@ def test_filter_with_function_returning_non_bool_raises_match_error(self) -> Non
...
+        self.assertEqual(
+            self._run(
+                """
+        range 3
+        """
+            ),
+            List([Int(0), Int(1), Int(2)]),
+        )
+
+    def test_range_non_int_raises_type_error(self) -> None:
```
Comment (tekknolagi) [inline scrapscript.py]: can you say `...with_non_int` (here and elsewhere?)

File: scrapscript.py [hunk 2]
```diff
@@ -3279,6 +3279,262 @@ def test_filter_with_function_returning_non_bool_raises_match_error(self) -> Non
...
+        self.assertEqual(
+            self._run(
+                """
+        foldr (x -> a -> a + x) 0 [1, 2, 3]
+        """
+            ),
+            Int(6),
+        )
+
+    def test_foldr_on_empty_list(self) -> None:
```
Comment (tekknolagi) [inline scrapscript.py]: `...returns_empty_list`

**Review by tekknolagi — APPROVED** [review]
In general: `test_<function>_with_<condition>_<does_something>`

lgtm

**Comment by tekknolagi** [comment]
Another alternative, if you are concerned about growth of prelude, is to work together to get scrap loading functionality working and then we can publish this somewhere

**Comment by gregorybchris** [comment]
> Another alternative, if you are concerned about growth of prelude, is to work together to get scrap loading functionality working and then we can publish this somewhere

I'm personally not concerned at all about the absolute size of the prelude. If the prelude is large I think that's ok. What I would be concerned about is 1. Adding something that's not standard in most programming languages 2. Adding something with inconsistent/confusing naming (e.g. fold vs reduce is potentially worth a few seconds of thought before merging)

I also think it'd be easy to throw them behind a CLI flag and/or in a new file, which imo addresses some of the potential simplicity concerns.

**Pull request merged**